data/src/test/java/org/apache/iceberg/TestMergingMetrics.java

```diff
@@ -0,0 +1,166 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.RandomGenericData;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.types.Types;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+public abstract class TestMergingMetrics<T> {
+
+  // all supported fields, except for UUID which is on deprecation path: see https://github.com/apache/iceberg/pull/1611
+  // as well as Types.TimeType and Types.TimestampType.withoutZone as both are not supported by Spark
+  protected static final Types.NestedField ID_FIELD = required(1, "id", Types.IntegerType.get());
+  protected static final Types.NestedField DATA_FIELD = optional(2, "data", Types.StringType.get());
+  protected static final Types.NestedField FLOAT_FIELD = required(3, "float", Types.FloatType.get());
+  protected static final Types.NestedField DOUBLE_FIELD = optional(4, "double", Types.DoubleType.get());
+  protected static final Types.NestedField DECIMAL_FIELD = optional(5, "decimal", Types.DecimalType.of(5, 3));
+  protected static final Types.NestedField FIXED_FIELD = optional(7, "fixed", Types.FixedType.ofLength(4));
+  protected static final Types.NestedField BINARY_FIELD = optional(8, "binary", Types.BinaryType.get());
+  protected static final Types.NestedField FLOAT_LIST = optional(9, "floatlist",
+      Types.ListType.ofRequired(10, Types.FloatType.get()));
+  protected static final Types.NestedField LONG_FIELD = optional(11, "long", Types.LongType.get());
+
+  protected static final Types.NestedField MAP_FIELD_1 = optional(17, "map1",
+      Types.MapType.ofOptional(18, 19, Types.FloatType.get(), Types.StringType.get())
+  );
+  protected static final Types.NestedField MAP_FIELD_2 = optional(20, "map2",
+      Types.MapType.ofOptional(21, 22, Types.IntegerType.get(), Types.DoubleType.get())
+  );
+  protected static final Types.NestedField STRUCT_FIELD = optional(23, "structField", Types.StructType.of(
+      required(24, "booleanField", Types.BooleanType.get()),
+      optional(25, "date", Types.DateType.get()),
+      optional(27, "timestamp", Types.TimestampType.withZone())
+  ));
+
+  private static final Set<Integer> IDS_WITH_ZERO_NAN_COUNT = ImmutableSet.of(1, 2, 5, 7, 8, 11, 24, 25, 27);
+
+  private static final Map<Types.NestedField, Integer> FIELDS_WITH_NAN_COUNT_TO_ID = ImmutableMap.of(
+      FLOAT_FIELD, 3, DOUBLE_FIELD, 4, FLOAT_LIST, 10, MAP_FIELD_1, 18, MAP_FIELD_2, 22
+  );
```

Comment:
I currently put it this way so that file types other than parquet could use it, however I've noticed that ORC reports metrics differently than parquet. I think ORC currently doesn't report anything for elements in list/map type, but report metrics for the list/map type themselves, which is completely opposite to what parquet does today. I think what parquet does is more useful, and wonder if people have opinion on if we want to record element metrics for ORC along with the new NaN counter. This is not a blocker for this specific pr, but the ORC implementation is on its way and I'd like to gather feedback sooner the better.

Reply:
I'm okay either way. Right now, we don't have any filter predicates that take advantage of nested stats, like `containsNull` or `containsNaN` for a column that is a `list<float>`. We will eventually want to support those predicates, but I don't think that we will add them in the short term since nothing really pushes them down for now. We can always add the stats for ORC when we start adding the predicates.

Reply:
Thank you for the explanation!